### ISSUE TITLE
Does the Node.js gRPC Service use "google.rpc.Status" for its status messages?

### ISSUE THREAD

**Issue by jmdobry** [grpc/grpc#9053]
**tldr;** I'm building a gRPC service and I'm struggling to find the best way to send an error to the client. When my server tries to send a nice `google.rpc.Status` message as the error, it fails.

Let's say I'm building a service that implements [`google.longrunning.Operations`](https://github.com/googleapis/googleapis/blob/master/google/longrunning/operations.proto), and I'm currently implementing the [`GetOperation`](https://github.com/googleapis/googleapis/blob/master/google/longrunning/operations.proto#L53) method. When the client does this: 
```js
operationsClient.getOperation({
  name: 'operations/abc123'
}, (err, operation) => {
  // ...
});
```
the server should respond with an [`Operation`](https://github.com/googleapis/googleapis/blob/master/google/longrunning/operations.proto#L82) message, which the client receives as the `operation` argument.

Now let's say the client requests an operation that doesn't exist:
```js
operationsClient.getOperation({
  name: 'operations/doesNotExist'
}, (err, operation) => {
  if (err) {
    console.error(err); // { code: 5, details: 'Operation operations/doesNotExist does not exist' }
  }
});
```

The problem with that error object is that it doesn't seem to match [`google.rpc.Status`](https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto), which one would normally be able to fill up with juicy error information according to [`google.rpc.error_details.proto`](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto). I'm limited by just a `details` field, which is a string, and the `metadata` field, which is key-value (<string, string|buffer>) only. 

This is what my server is trying to send as the error (`err` is an instance of `Error`):

```js
{
  code: grpc.status.NOT_FOUND,
  message: 'Operation operations/doesNotExist does not exist',
  details: [
    {
      typeUrl: 'types.googleapis.com/google.rpc.DebugInfo',
      value: {
        stackEntries: [
          'Error: Operation operations/doesNotExist does not exist',
          '    at Object.getOperation (/path/to/file.js:49:17)',
          // ...
        ],
        detail: err.message
      }
    },
    {
      typeUrl: 'types.googleapis.com/google.rpc.ResourceInfo',
      value: {
        resourceType: 'types.googleapis.com/google.longrunning.Operation',
        resourceName: 'operations/doesNotExist',
        description: 'Operation operations/doesNotExist does not exist'
      }
    }
  ]
}
```

Here it is with the objects of type `google.protobuf.Any` encoded:

```js
{
  "code": 5,
  "message": "Operation operations/doesNotExist does not exist",
  "details": [
    {
      "typeUrl": "types.googleapis.com/google.rpc.DebugInfo",
      "value": "eyJzdGFja0VudHJpZXMiOlsiRXJyb="
    },
    {
      "typeUrl": "types.googleapis.com/google.rpc.ResourceInfo",
      "value": "eyJyZXNvdXJjZVR5cGUiOiJ0eXBlc="
    }
  ]
}
```

It seems like the Node.js gRPC server should be able to send my `google.rpc.Status` message as the error. I'm looking at the [`handleError`](https://github.com/grpc/grpc/blob/master/src/node/src/server.js#L75) method in `server.js`, and it tries constructing its own status to send to the client. However, the status object it creates does not seem to match `google.rpc.Status`. Instead, the `status` created by `handleError` has its `details` field as a string, and it doesn't even have a `message` field.

I've considered using the `Metadata` class (`handleError` will put a `metadata` field on its `status` object), but it's only key-value pairs. I'd have to smash my nice Status message into key-values pairs and deconstruct that on the client.

How can I send a rich `google.rpc.Status` to the client as the error? Can Node.js gRPC be made to us `google.rpc.Status` as its status message format?

**Comment by jmdobry** [comment]
Possibly related to grpc/grpc.github.io#368

**Comment by nevi-me** [comment]
Hey @jmdobry it so far seems like the error message design is similar to `google.rpc.Status`, but there are some differences that I've noted.

* in your example, you included `message` and `details`, the Node.js code seems to overwrite `message` with `details` if `code` and `details` are both present.
* the above means that you can't pass `Any` as `details` because `message` is supposed to be a `string`.

The above is based on looking at [http://www.grpc.io/grpc/node/src_server.js.html#line75](http://www.grpc.io/grpc/node/src_server.js.html#line75)

As you mentioned about the `Metadata` being KV only, you can still pass a buffer as value (the key should end with `-bin`), but I don't know how practical that would be. I'll try manually creating a Protobuf message with Protobuf.js, and try send that as Metadata. I quickly tried this using a Node server and an Android client.

**Comment by nevi-me** [comment]
I'm not familiar with the HTTP2 spec, so not sure what headers should be used to send. However, I've just successfully tested sending a protobuf message as the `Metadata` of my request, and parsed it from a `ByteArray` on Android.

What you could do then is pass `details-bin` as your error message, as an encoded `Any` message. Clients would then check for the presence of `details-bin` in the headers, and then decode them.

It's not elegant as it leaves implementation details to the client, but I think I'm going to follow that approach.

**Comment by jmdobry** [comment]
Here is what I came up with (though I omit how I encoded/decoded the messages of type `google.protobuf.Any`):

The server has a Status message :

```js
{
  "code": 5,
  "message": "Operation operations/doesNotExist does not exist",
  "details": [
    {
      "typeUrl": "types.googleapis.com/google.rpc.DebugInfo",
      "value": "eyJzdGFja0VudHJpZXMiOlsiRXJyb="
    },
    {
      "typeUrl": "types.googleapis.com/google.rpc.ResourceInfo",
      "value": "eyJyZXNvdXJjZVR5cGUiOiJ0eXBlc="
    }
  ]
}
```

I prepare it to be passed to the client:

```js
const error = {
  code: status.code || grpc.status.INTERNAL,
  details: status.message || http.STATUS_CODES['500']
};

if (Array.isArray(status.details)) {
  const metadata = new grpc.Metadata();
  status.details.forEach((detail, i) => {
    if (i) {
      metadata.add('detail-bin', Buffer.from(JSON.stringify(detail)));
    } else {
      metadata.set('detail-bin', Buffer.from(JSON.stringify(detail)));
    }
  });
  error.metadata = metadata;
}

// Send the error status message to the client
call(error);
```

And then on the client:

```js
if (err.metadata) {
  const details = err.metadata.get('detail-bin');
  if (Array.isArray(details)) {
    err.details = details.map((detail) => JSON.parse(detail.toString()));
  }
  err.metadata.remove('detail-bin');
  if (!Object.keys(err.metadata.getMap()).length) {
    delete err.metadata;
  }
}
```

and now the client has the same representation that the server sent:

```js
{
  "code": 5,
  "message": "Operation operations/doesNotExist does not exist",
  "details": [
    {
      "typeUrl": "types.googleapis.com/google.rpc.DebugInfo",
      "value": "eyJzdGFja0VudHJpZXMiOlsiRXJyb="
    },
    {
      "typeUrl": "types.googleapis.com/google.rpc.ResourceInfo",
      "value": "eyJyZXNvdXJjZVR5cGUiOiJ0eXBlc="
    }
  ]
}
```

**Comment by nevi-me** [comment]
Great stuff! I think passing the data as metadata is the only solution.
